### ISSUE TITLE
Create debug version of SvarCOM

### ISSUE THREAD

**Issue by boeckmann** [SvarDOS/bugz#148]
Follow-up of https://github.com/SvarDOS/bugz/issues/147#issuecomment-2565575298 and https://github.com/SvarDOS/bugz/issues/147#issuecomment-2565630254.

**Comment by boeckmann** [comment]
@ecm-pushbx made a comment about the details when dealing with flat-formated (.COM) larger than 64k: https://github.com/SvarDOS/bugz/issues/147#issuecomment-2565685011

**Comment by boeckmann** [comment]
I was able to run the 108K .COM binary containing the debug info under the EDR kernel with the Open Watcom debugger. Source-level debugging (triggering breakpoints, inspecting locals etc.) seems to work.

However, mid-term I think it would be better for debugging to also have the possibility to build a .EXE version of SvarCOM. This should also work under DosBox and MS-DOS. Will be tricky to get this running with CS = DS = SS = PSP. As I am currently able to debug I will leave this for some time later when I have time to think about it.

**Comment by boeckmann** [comment]
@ecm-pushbx hints may be helpful in creating the .EXE:

> The debugger's own build process creates the executable image as a ".BIG" file. It then wraps that image in either an inicomp stage, directly as the MZ image of the lDOS iniload stage, or [within an "MZ shim" stage](https://hg.pushbx.org/ecm/ldebug/file/e2fafd3b7668/source/mzshim.asm). Example use is [in the mak script](https://hg.pushbx.org/ecm/ldebug/file/e2fafd3b7668/source/mak.sh#l466). Importantly it uses the auto stack feature which, alike the /E switch to lDebug, sets up the stack in an SS != PSP and SS != CS. The default is to use a CS == PSP, which involves an MZ header field for init CS of -16. With an init IP of 256 this can be used to load big images in a way that is similar to running a smaller flat-format / .COM file, except for the SS != PSP and there not being a zero word as a return address on the initial stack.

**Comment by boeckmann** [comment]
I created a simple [.COM to .EXE converter](https://github.com/boeckmann/com2exe/blob/main/com2exe.c). The >64K COMMAND.COM debug binary converted with it runs under DosBox. But curiously now OpenWatcom WD has a strange problem with it: It runs it on WD launch, giving me a command prompt, instead of waiting in the source level debugging interface. It may be that the debug info contains some relative offsets or so into the file. That would be a show-stopper for this way of doing it.

**Comment by ecm-pushbx** [comment]
I think this line is wrong, you want modulo 512 not shift right by 9: https://github.com/boeckmann/com2exe/blob/4cafdf4f81b25fcc2213dea3e3411189abb63864/com2exe.c#L99

**Comment by ecm-pushbx** [comment]
Also your pages amount field is off by one, you want to count the entire size (image + header), not just the image alone: https://hg.pushbx.org/ecm/ldebug/file/e2fafd3b7668/source/mzshim.asm#l51

**Comment by ecm-pushbx** [comment]
You should also set minimum allocation correctly (not zero) in case the input file is <= FF00h bytes long.

**Comment by boeckmann** [comment]
> I think this line is wrong, you want modulo 512 not shift right by 9: https://github.com/boeckmann/com2exe/blob/4cafdf4f81b25fcc2213dea3e3411189abb63864/com2exe.c#L99

Ouch. That one hurts...

**Comment by boeckmann** [comment]
> Also your pages amount field is off by one, you want to count the entire size (image + header), not just the image alone: https://hg.pushbx.org/ecm/ldebug/file/e2fafd3b7668/source/mzshim.asm#l51

Thanks for clarification. I questioned myself if the header has to be included, but under https://wiki.osdev.org/MZ etc. found no difinitive answer to that.

**Comment by boeckmann** [comment]
> You should also set minimum allocation correctly (not zero) in case the input file is <= FF00h bytes long.

Good point. As I unconditionally set SP=0xfffe it should be made sure that the .COM has at least 64k memory to work with.

**Comment by boeckmann** [comment]
Updated according to your comments:
https://github.com/boeckmann/com2exe/commit/7d670b6e71e25908e27cbb045497034f77ab727a

**Comment by boeckmann** [comment]
Looks like trying to symbolically debug the EXEfied .COM file is a dead end. While the .EXE file itself runs without problems the Watcom Debugger is confused about the symbol offsets. The symbol segments are all off by 10h frames. Probably because the debugger assumes that CS points to after the PSP, or it is encoded directly in the symbol info, I dunno.

Instead of trying to fix this I think it would be more appropriate to make SvarCOM buildable as an .EXE file. Or even always build an EXE and name it command.com nevertheless.

**Comment by boeckmann** [comment]
I was able to create a preliminary .EXE version of SvarCOM by adapting startup.asm, the makefile and changing some C code, especially replacing hard-coded near pointers into the PSP with something like `MK_FP(_psp,0x12)`. As DS != PSP when building an .EXE, more such changes would have to be performed.

Watcom Debugger is able to handle the .EXE file created this way.

@mateuszviste would it be ok to you if the required changes get included, so that SvarCOM can be built either as a .COM or a .EXE binary?

**Comment by mateuszviste** [comment]
While I do not like the idea of SvarCOM being an EXE file, I do not have objections about an extra EXE target for debugging purposes. I'd really prefer this not to increase the size of the COM variant, so probably some ifdefs would be in order.

**Comment by boeckmann** [comment]
This Watcom Debugger stuff is causing me pain. WD freaks out on a normally generated .EXE (apart from the missing standard lib):

<img width="667" alt="Bildschirmfoto 2025-01-02 um 02 13 19" src="https://github.com/user-attachments/assets/f4f9e86f-4aaf-4e74-b782-e20270f99838" />

`char *cmdline_g` is a global variable that lives at `1c31:3ec0`. So far, so good. `char *cmdlinelen` is a local _near_ pointer to `3ec0`, meaning it aliases `cmdline_g`, like it should. Correct so far. However, only `cmdline_g` is shown correct in the debugger. When I inspect the cmdlinelen, WD assumes a totally wrong segment, namely `20a6` instead of `1c31`. No idea where this comes from. For sure not from a segment register, as these point to the correct segment `1c31`. The code also runs perfectly fine. If that debugging stuff continues to turn me down I will resort to using a combination of printf, int 3 and lDebug :D

**Comment by boeckmann** [comment]
When including the Watcom standard library it works as intended. After an hour trial and error I found that the Watcom Debugger expects a symbol `__nullarea` as part of the `_NULL` segment to be defined. Seems that it uses this to some segment arithmetic. I adjusted the WMINCRT startup code accordingly, which finally made it work. Tomorrow I will clean up the mess I introduced to the files while guessing what went wrong...

**Comment by mateuszviste** [comment]
Maybe best would be for the debug-exe version to include the standard libc?

**Comment by boeckmann** [comment]
Initial debug capability is now implemented by [rev 2315](https://github.com/SvarDOS/core/commit/e3e0d0a2725dad2c1bb795d444696b89ac98805d).

A debuggable MZ executable may be built via:
```
wmake clean
wmake DEBUG=1
```

Otherwise, it is built as a flat .COM binary, like before. Non-debug version should not have increased much in size. Most of the stuff is encapsulated by IFDEF conditionals. I had to change a few pointers to FAR, especially the PSP handling stuff.

I tried to resemble the .COM behaviour as much as possible. Even reproduced the (dangerous) fact that the DTA and command line tail share a single buffer. In the debug executable, this 80h byte buffer is relocated from the PSP to some buffer in the BSS segment, as many DTA operations are coded via near pointers (which is ok considering the performance benefits).

I implemented a simple null pointer write check. Will extend this to the first 256 bytes of the executable, to be able to catch "wild" writes to the PSP etc. via near pointers I missed...

**Comment by mateuszviste** [comment]
The non-DEBUG (COM) version of the "EXE" branch appears to be 3% bigger (ca. 1.5K) than the current trunk version. I will try to find time to look why, maybe there are still some ifdefs that could bring this down.

The new WMINCRT startup.obj went from 540 bytes to 774 bytes. It's only some 200+ bytes, so not really meaningful for SvarCOM, but could be more important for some other tools that would use it in the future. WMINCRT allows to write C programs that are almost as compact as asm equivalents, it would be really nice to keep WMINCRT as barebone as possible (maybe commenting out the debug-related stuff at assembly time?)

**Comment by boeckmann** [comment]
Do not look any further. I forgot to include the _-ox_ code optimization switch in the updated makefile. Now it is down to 47676 bytes :)

**Comment by boeckmann** [comment]
The increased size of the .OBJ are mainly the additionally defined segments etc. Most of them are empty. This added .OBJ size does not transfer 1:1 to the .COM executable size.

**Comment by mateuszviste** [comment]
> The increased size of the .OBJ are mainly the additionally defined segments etc. Most of them are empty. This added .OBJ size does not transfer 1:1 to the .COM executable size.

The hello world example from the WMINCRT directory on svn compiles to 134 bytes when using the old WMINCRT. When using the new one from the SvarCOM EXE branch, the size of HELLO.COM goes to 386 bytes. So it's actually more than the size difference between the old and new startup.obj. Linker magic, I guess.

It is debatable whether or not an increase of 252 bytes is really bothersome for any practical purposes, but it definitely takes some sex-appeal away from WMINCRT. :-)

**Comment by boeckmann** [comment]
You are right. I did not test it with WMINCRT until now. The increase results from the additionally defined segments. I put them into IF conditions in rev 2323. Now the hello world is down to 156 bytes. Did not have such an effect on command.com though.

**Comment by boeckmann** [comment]
Size of an .EXE version of the wmincrt hello example is 244 bytes. Not that bad :)

**Comment by boeckmann** [comment]
The increased size of the .COM was caused by the `ORG 100h` directive giving 256 extra bytes, because there was the BEG_TEXT segement defined in front _TEXT.

**Comment by boeckmann** [comment]
I cherry-picked the single missing commit r2318 (volume id string fix) into branch/exe. It should now be on the same level as trunk. Size problem is also fixed. So could be merged back into trunk, if nothing speaks against it.

Independent of if / how this is merged, 2025.0 should be released as soon as possible, as people start to redistribute 2024.8 in custom made images like: https://github.com/lproven/usb-dos

**Comment by mateuszviste** [comment]
Would you mind creating a new "release" (at least a svn tag) of WMINCRT? It would be clearer then what SvarCOM uses exactly.
It would also probably be good to put some files on the [WMINCRT web page](http://svardos.org/wmincrt/) so humanity can start enjoying this neat piece of work :)

**Comment by mateuszviste** [comment]
trunk is up to date with the exe branch now, so I think it should be safe to delete the branch.

**Comment by boeckmann** [comment]
> Would you mind creating a new "release" (at least a svn tag) of WMINCRT? It would be clearer then what SvarCOM uses exactly.

Sure, we can do this. SvarCOM `crt.h` should be included it in wmincrt as `wmincrt.h` before tagging. Also for an official wincrt release it would be best not to utilize the `crt_temp_cmdline` and `crt_temp_dta`, first they overlap (to make code changes to svarcom minimal) and second I would rather try to support the standard C like `main(int argc, char (argv[])`. But not quite sure yet. Would mean a divergence of the SvarCOM crt with wmincrt. For me that would not be that of a problem though...

**Comment by boeckmann** [comment]
> trunk is up to date with the exe branch now, so I think it should be safe to delete the branch.

Perfect. How do I delete it? Just `svn rm`?

**Comment by mateuszviste** [comment]
> Sure, we can do this. SvarCOM crt.h should be included it in wmincrt as wmincrt.h before tagging. Also for an official wincrt release it would be best not to utilize the crt_temp_cmdline and crt_temp_dta, first they overlap (to make code changes to svarcom minimal) and second I would rather try to support the standard C like main(int argc, char (argv[]).

I think at this point it would be best for SvarCOM to simply use a proper dta buffer. I will look into that, probably during the weekend. Then SvarCOM could use the vanilla WMINCRT. A pointer to the cmdline area would still be helpful, though.

Dealing with argv[] business inside WMINCRT sounds somewhat expensive. But maybe it could be automatically disabled when the program uses a main(void) invocation?

> > trunk is up to date with the exe branch now, so I think it should be safe to delete the branch.
> 
> Perfect. How do I delete it? Just `svn rm`?

yes (unless you see anything missing in trunk of course, but I double checked so it should all be there).

**Comment by boeckmann** [comment]
> I think at this point it would be best for SvarCOM to simply use a proper dta buffer. I will look into that, probably during the weekend.

Great! Then I modify WMINCRT this weekend to not do the DTA stuff anymore but leave the cmdline stuff as it is now, only rename the variable to crt_cmdline. As is is not shared anymore with the DTA, it is not a temp anymore ;)

Supporting the main(int,char**) thing is then something for a future release.
